File: agileforge-frontend/src/app/core/services/gamification.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { GamificationProfile, Badge, LeaderboardEntry, Achievement } from '../models/ecosystem.model';

@Injectable({ providedIn: 'root' })
export class GamificationService {
  private http = inject(HttpClient);
  private baseUrl = '/api/gamification';

  getProfile(userId: string): Observable<GamificationProfile> {
    return this.http.get<GamificationProfile>(`${this.baseUrl}/profile/${userId}`);
  }

  getAvailableBadges(): Observable<Badge[]> {
    return this.http.get<Badge[]>(`${this.baseUrl}/badges`);
  }

  getLeaderboard(organizationId: string, period: string = 'WEEKLY'): Observable<LeaderboardEntry[]> {
    return this.http.get<LeaderboardEntry[]>(`${this.baseUrl}/leaderboard/${organizationId}`, { params: { period } });
  }

  optInLeaderboard(userId: string, organizationId: string, optIn: boolean): Observable<void> {
    return this.http.post<void>(`${this.baseUrl}/leaderboard/opt-in`, { userId, organizationId, optIn });
  }

  getTeamAchievements(projectId: string): Observable<Achievement[]> {
    return this.http.get<Achievement[]>(`${this.baseUrl}/achievements/${projectId}`);
  }

  recordActivity(userId: string, activityType: string, metadata?: Record<string, unknown>): Observable<void> {
    return this.http.post<void>(`${this.baseUrl}/activity`, { userId, activityType, metadata });
  }
}

File: agileforge-backend/src/main/java/com/agileforge/application/service/GamificationService.java
package com.agileforge.application.service;

import com.agileforge.domain.model.*;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.*;

@Service
public class GamificationService {

    public record UserProfile(UUID userId, List<UserBadge> badges, List<UserStreak> streaks,
                              List<UserXp> xpByDomain, int totalXp, int level) {}

    public record LeaderboardEntry(UUID userId, String userName, int points, int rank, int badgeCount) {}

    public record Achievement(UUID id, String name, String description, String type, List<UUID> participants) {}

    private final List<Badge> allBadges = new ArrayList<>();

    public UserProfile getUserProfile(UUID userId) {
        List<UserBadge> badges = getUserBadges(userId);
        List<UserStreak> streaks = getUserStreaks(userId);
        List<UserXp> xp = getUserXp(userId);
        int totalXp = xp.stream().mapToInt(UserXp::getXpPoints).sum();
        int level = calculateLevel(totalXp);
        return new UserProfile(userId, badges, streaks, xp, totalXp, level);
    }

    public List<UserBadge> getUserBadges(UUID userId) {
        return List.of();
    }

    public List<UserStreak> getUserStreaks(UUID userId) {
        return List.of();
    }

    public List<UserXp> getUserXp(UUID userId) {
        return List.of();
    }

    public List<Badge> getAvailableBadges() {
        return allBadges;
    }

    public void recordActivity(UUID userId, String activityType, Map<String, Object> metadata) {
        updateStreak(userId, activityType);
        awardXp(userId, activityType, metadata);
        checkBadgeEligibility(userId, activityType);
    }

    public void updateStreak(UUID userId, String activityType) {
        // Update streak count for delivery activities
    }

    public void awardXp(UUID userId, String activityType, Map<String, Object> metadata) {
        // Award XP based on activity type and domain
    }

    public void checkBadgeEligibility(UUID userId, String activityType) {
        // Check if user qualifies for any new badges
    }

    public List<LeaderboardEntry> getLeaderboard(UUID organizationId, String period) {
        return List.of();
    }

    public void optInLeaderboard(UUID userId, UUID organizationId, boolean optIn) {
        // Toggle leaderboard visibility
    }

    public List<Achievement> getTeamAchievements(UUID projectId) {
        return List.of();
    }

    public void celebrateMilestone(UUID projectId, String milestoneName, List<UUID> participants) {
        // Record team achievement
    }

    private int calculateLevel(int totalXp) {
        if (totalXp < 100) return 1;
        if (totalXp < 300) return 2;
        if (totalXp < 600) return 3;
        if (totalXp < 1000) return 4;
        if (totalXp < 1500) return 5;
        if (totalXp < 2500) return 6;
        if (totalXp < 4000) return 7;
        if (totalXp < 6000) return 8;
        if (totalXp < 9000) return 9;
        return 10;
    }
}
